Add MINIMUM_STEPPER_PULSE ability to babystepping

Sebastianv650 opened the pull request:
For whatever reason babystepping isn't respecting MINIMUM_STEPPER_PULSE, see #6072. This PR should fix this. But it's **untested** because I'm not using babystepping.
@matrix3dmachine please test and report if it works.

This also adds protection against other (especialy stepper) ISRs, which gets even more important when the runtime of a babystep is increased due to MINIMUM_STEPPER_PULSE.
This eleminates the basic need of #5949 up to some point.

Commit:
Add MINIMUM_STEPPER_PULSE ability to babystepping

Sebastianv650 commented:
CYCLES_EATEN_BY_BABYSTEP is taken from the extruder advance ISR, I didn't measure them. I just guessed that it should be compareable. If somebody can provide a better number, feel free to change it!

matrix3dmachine commented:
Just copied the code into stepper.cpp.

Get lots of compile errors:
```
stepper.cpp:1243: error: '#' is not followed by a macro parameter
    #define BABYSTEP_AXIS(axis, AXIS, INVERT) { \ 
                                            ^
sketch\stepper.cpp: In static member function 'static void Stepper::babystep(AxisEnum, bool)':
stepper.cpp:1269: error: 'x' was not declared in this scope
          BABYSTEP_AXIS(x, X, false); 
                        ^
stepper.cpp:1269: error: 'X' was not declared in this scope
          BABYSTEP_AXIS(x, X, false); 
                           ^
stepper.cpp:1269: error: 'BABYSTEP_AXIS' was not declared in this scope
          BABYSTEP_AXIS(x, X, false); 
                                   ^
stepper.cpp:1274: error: 'y' was not declared in this scope
          BABYSTEP_AXIS(y, Y, false); 
                        ^
stepper.cpp:1274: error: 'Y' was not declared in this scope
          BABYSTEP_AXIS(y, Y, false); 
                           ^
stepper.cpp:1284: error: 'z' was not declared in this scope
            BABYSTEP_AXIS(z, Z, BABYSTEP_INVERT_Z); 
                          ^
stepper.cpp:1284: error: 'Z' was not declared in this scope
            BABYSTEP_AXIS(z, Z, BABYSTEP_INVERT_Z); 
                             ^
exit status 1
'#' is not followed by a macro parameter
```

Sebastianv650 commented:
Just learned that
-Travis CI pass is not always finding all errors
-You can't use #if inside a macro

I will do it again, this time with more detail checking ;)

Sebastianv650 commented:
Most likely @thinkyhead will not like my coding style, but this is now the version which I got to compile.
I'm bad at macros and I don't like them (it's hard to read them for me), hard to say which is the cause of the other ;)
So he might provide a nicer version with a compact macro in it, but up to then you can test and use this one.

Commit:
Protect Babystepping against other ISRs

Especialy against stepper ISR.
This is even more important when a minimum pulse width is set,
increasing the runtime of a babystep.

thinkyhead commented:
> Travis CI pass is not always finding all errors

Well, you have to add a test to `.travis.yml` or Travis won't try to compile with the relevant options. Check out that file, if you haven't already. It used to be unreadable, but I wrote some helpful scripts to set configuration file options from the command-line and now it's much more obvious what's happening in there.

thinkyhead commented:
This looks very good. I see it has the `cli`/`sei` added as well. That should be good enough to prevent any conflicts with the stepper ISR, and otherwise I think the Temperature ISR isn't causing any more mayhem. (But we'll keep an eye on that.)

thinkyhead commented on Marlin/stepper.cpp:
```diff
@@ -1231,40 +1231,73 @@ void Stepper::report_positions() {
+    static uint8_t old_pin;
+    #if STEP_PULSE_CYCLES > CYCLES_EATEN_BY_BABYSTEP
+      static uint32_t pulse_start;
+    #endif
```
Actually, let's change this. I don't think there's any advantage to having `pulse_start` as a `static`, and the same with `old_pin`. They could both benefit from being able to be in registers, if the compiler is able to do so.

matrix3dmachine commented:
Hi Scott,
I've tested the code that Sebastian wrote for me. 

It works perfect! 

That's to all you guys for your help!

Can this be included into the next RC9 release?

Mike

> On Mar 22, 2017, at 7:43 AM, Scott Lahteine <notifications@github.com> wrote:
> 
> This looks very good. I see it has the cli/sei added as well. That should be good enough to prevent any conflicts with the stepper ISR, and otherwise I think the Temperature ISR isn't causing any more mayhem. (But we'll keep an eye on that.)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

thinkyhead commented:
> Can this be included into the next RC9 release?

Most definitely!
